quinn-proto/src/connection/mod.rs

```diff
@@ -947,6 +950,29 @@ where
     ///
     /// Returns the number of bytes successfully written.
     pub fn write(&mut self, stream: StreamId, data: &[u8]) -> Result<usize, WriteError> {
+        let mut source = ByteSlice::from_slice(data);
+        self.write_source(stream, &mut source)?;
+        Ok(source.bytes_consumed())
+    }
+
+    /// Send data on the given stream
+    ///
+    /// Returns the number of bytes successfully written.
+    pub fn write_chunks(
+        &mut self,
+        stream: StreamId,
+        data: &mut [Bytes],
+    ) -> Result<(usize, bool), WriteError> {
```

Comment:
Why is it useful to return the `partial_chunk_consumed()` value here? Presumably the caller can just forget about/recycle the consumed chunks, and the partially consumed chunk has been split, such that it's equivalent to an unconsumed chunk, isn't it?

Reply:
The caller needs to be aware that this buffer had been mutated. We can't claim it as fully consumed, since the caller needs to submit the leftovers again in another write call. For fully consumed buffers, the caller will just drop/skip them after the call. For the partially consumed one this isn't possible.

An alternative can be to report the number of mutated buffers, including the last one. But this requires the caller to check the buffers and see if one wasn't fully consumed.

Reply:
Sorry, perhaps I'm being stupid -- why does the caller need to be aware that the buffer has been mutated? We could just mark it as unconsumed.

Reply:
One example that I have in code around it is a timeout wrapper around the stream.

If 0 chunks have been consumed, it is equivalent to `Pending` and means the timer needs to continue to run. If a partial chunk is consumed this means the stream is still active and we have to reset the timer. 

Something like that could probably also be done by comparing the size of the first buffer before and after the write, but I kind of liked the partial signal. I guessed there might also be other use-cases where it's important whether e.g. the 3rd buffer was partially consumed or not, but I don't have a concrete example for it. The code I have for this will retry any non-fully consumed buffers again, so there is no special handling required.

So I guess I would also be fine with only reporting the amount of fully consumed buffers, or by reporting the amount of consumed buffers plus consumed bytes

Reply:
Yes, I think reporting fully consumed buffers + consumed bytes seems like the right approach here. That should easily satisfy the timeout wrapper use case at least (I agree that comparing the size of the first buffer is a little unwieldy).